internal/staging_lockfile/staging_lockfile.go

```diff
@@ -0,0 +1,173 @@
+package staging_lockfile
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/containers/storage/internal/rawfilelock"
+)
+
+// StagingLockFile represents a file lock used to coordinate access to staging areas.
+// Typical usage is via CreateAndLock (for a new file) or TryLockExisting (for an existing file),
+// both of which return a StagingLockFile that must eventually be released with UnlockAndDelete.
+// This ensures that access to the staging file is properly synchronized both within and across processes.
+//
+// WARNING: This struct MUST NOT be created manually. Use the provided helper functions instead.
+type StagingLockFile struct {
```

Comment:
Something like
```suggestion
type StagingLockFile struct {
    // Locking invariant: If stagingLockFileLock is not locked, a StagingLockFile for a particular
    // path exists iff the current process currently owns the lock for that file, and it is recorded in stagingLockFiles.
```

to help future readers understand the design?